api/enki.py

```diff
@@ -0,0 +1,556 @@
+        )
+        self.batch_size = batch_size
+        self.api = api or EnkiAPI.from_environment(self._db)
+        self.bibliographic_coverage_provider = (
+            EnkiBibliographicCoverageProvider(self._db, enki_api=api)
+        )
+
+    def run_once(self, start, cutoff):
+        # Give us five minutes of overlap because it's very important
+        # we don't miss anything.
+        since = start-self.FIVE_MINUTES
+        id_start = 0
+        while True:
+            if availability.status_code != 200:
+                self.log.error(
+                "Could not contact Enki server for content availability. Status: %d",
+                availability.status_code
+            )
+            content = availability.content
+            count = 0
+            for bibliographic, circulation in BibliographicParser().process_all(content):
+                self.process_book(bibliographic, circulation)
+                count += 1
+                if count == 0:
+                    self._db.commit()
+            id_start += self.batch_size
+
+    def process_book(self, bibliographic, availability):
+
+        license_pool, new_license_pool = availability.license_pool(self._db)
+        edition, new_edition = bibliographic.edition(self._db)
+        license_pool.edition = edition
+        policy = ReplacementPolicy(
+            identifiers=False,
+            subjects=True,
+            contributions=True,
+            formats=True,
+        )
+        availability.apply(
+            pool=license_pool,
+            replace=policy,
+        )
+        if new_edition:
+            bibliographic.apply(edition, replace=policy)
+
+        if new_license_pool or new_edition:
+            # At this point we have done work equivalent to that done by
+            # the EnkiBibliographicCoverageProvider. Register that the
+            # work has been done so we don't have to do it again.
+            identifier = edition.primary_identifier
+            self.bibliographic_coverage_provider.handle_success(identifier)
+            self.bibliographic_coverage_provider.add_coverage_record_for(
+                identifier
+            )
+
+        return edition, license_pool
+
+class EnkiCollectionReaper(IdentifierSweepMonitor):
+    """Check for books that are in the local collection but have left the Enki collection."""
+    def __init__(self, _db, api=None, interval_seconds=3600*4):
+        super(EnkiCollectionReaper, self).__init__(_db, "Enki Collection Reaper", interval_seconds)
+        self._db = _db
+        if not api:
+            api = EnkiAPI.from_environment(_db)
+        self.api = api
+        self.data_source = DataSource.lookup(self._db, DataSource.ENKI)
+
+    def run(self):
+        super(EnkiCollectionReaper, self).run()
+
+    def identifier_query(self):
+        return self._db.query(Identifier).filter(
+            Identifier.type==Identifier.ENKI_ID)
+
+    def process_batch(self, identifiers):
```

Comment:
Since you're asking about one item at a time, I recommend copying `OverdriveCollectionReaper`. It'll let you get rid of a lot of code and also make this code easier to port to the 2.0 monitoring system. All this batch handling code is for Bibliotheca, where you can ask about 25 items at a time.